### DISCUSSION TITLE
Does Shadow support tokio?

### DISCUSSION THREAD

**Discussion by haxxpop** [shadow/shadow#3241] [Q&A]
I saw that there are no tests for tokio. I'm curious if it's supported. If it is, do you have a plan to add the tests for tokio?

**Comment by stevenengler** [comment]
This repository doesn't have any tests that use Tokio, but the Tor Project runs tests of [Arti](https://gitlab.torproject.org/tpo/core/arti) in Shadow. So Tokio should work fine in most use-cases.

Shadow doesn't "officially" support Tokio since I don't think anyone has the time to make a long-term commitment to making sure future versions of Tokio continue to work in Shadow, and I wouldn't be surprised if there are some current Tokio features that don't work correctly in Shadow. But I think it would be helpful to have tests that use Tokio to make sure Shadow doesn't accidentally break any existing Tokio support.

For tests, my personal opinion is that it's usually best to focus on testing applications rather than testing libraries since applications usually have a more-stable interface than libraries. So if we wanted to test Tokio, it's nicer to test an application that uses Tokio (and ideally make this into an example in `examples/apps/` as well). But I don't know what anyone else's thoughts are on this.

**Reply by haxxpop** [reply]
I have a good news. I'm adding some tests for rust-libp2p which uses tokio.

**Reply by robgjansen** [reply]
I just wanted to say I agree with @stevenengler thoughts too.

My main concern is that any added tests should be easy to maintain, which I think we could guarantee by pinning a version of tokio in the test environment. I suppose if some future tests were added that required a newer version of tokio than the one we have pinned, then it would be up to whoever wants to add the new tests to go and fix any older tokio code that breaks too.
